File: Review-04-Unit-Testing/Example-3-Python/tests/examples/test_board.py
from hamcrest import *
import unittest

from examples.board import Board

import copy


class TestBoard(unittest.TestCase):
    """
    1 - Does this piece of code perform the operations
        it was designed to perform?

    2 - Does this piece of code do something it was not
        designed to perform?

    1 Test per mutator
    """
    expected_empty_str = "1|2|3\n4|5|6\n7|8|9"

    def setUp(self):

        self.a_board = Board()

    def test_default_constructor(self):

        for expected_char in range(1, 10):
            assert_that(self.a_board.get_cell(expected_char),
                        is_(str(expected_char)))

        assert_that(str(self.a_board), equal_to(TestBoard.expected_empty_str))
        self.assertFalse(self.a_board.is_full())

        retrieved = self.a_board.get_3_cells(1, 2, 3)
        expected = ((0, '1'), (1, '2'), (2, '3'))

        assert_that(retrieved[0], equal_to(expected[0]))
        assert_that(retrieved[1], equal_to(expected[1]))
        assert_that(retrieved[2], equal_to(expected[2]))

    def test_set_cell(self):

        self.a_board.set_cell(1, 'X')
        self.a_board.set_cell(9, 'O')

        assert_that(self.a_board.get_cell(1), is_('X'))
        assert_that(self.a_board.get_cell(9), is_('O'))

        retrieved = self.a_board.get_3_cells(1, 5, 9)
        expected = ((0, 'X'), (4, '5'), (8, 'O'))

        assert_that(retrieved[0], equal_to(expected[0]))
        assert_that(retrieved[1], equal_to(expected[1]))
        assert_that(retrieved[2], equal_to(expected[2]))

        assert_that(str(self.a_board),
                    is_not(equal_to(TestBoard.expected_empty_str)))
        assert_that(str(self.a_board), equal_to("X|2|3\n4|5|6\n7|8|O"))

        self.assertFalse(self.a_board.is_full())

File: Review-04-Java-Unit-Testing/Example-1/src/test/java/edu/odu/cs/cs330/examples/TestBoard.java
package edu.odu.cs.cs330.examples;

import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;

import static org.junit.jupiter.api.Assertions.*;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
// import org.hamcrest.core.IsNull;

/**
 * 1 - Does this piece of code perform the operations
 *     it was designed to perform?
 *
 * 2 - Does this piece of code do something it was not
 *     designed to perform?
 *
 * 1 Test per mutator
 */
@TestMethodOrder(MethodOrderer.MethodName.class)
public class TestBoard
{
    final String expectedEmptyStr = "1|2|3\n4|5|6\n7|8|9\n";

    Board aBoard;

    @BeforeEach
    public void setUp()
    {
        aBoard = new Board();
    }

    @Test
    public void testDefaultConstructor()
    {
        for (int i = 1; i <= 9; i++) {
            char expectedChar = Character.forDigit(i, 10);
            assertThat(aBoard.getCell(i), is(expectedChar));
        }

        assertThat(aBoard.toString(), equalTo(expectedEmptyStr));
        assertFalse(aBoard.isFull());

        Board.Pair[] retrieved = aBoard.get3Cells(1, 2, 3);
        Board.Pair[] expected = {
            new Board.Pair(0, '1'),
            new Board.Pair(1, '2'),
            new Board.Pair(2, '3')
        };

        assertThat(retrieved[0], equalTo(expected[0]));
        assertThat(retrieved[1], equalTo(expected[1]));
        assertThat(retrieved[2], equalTo(expected[2]));
    }

    @Test
    public void testSetCell()
    {
        aBoard.setCell(1, 'X');
        aBoard.setCell(9, 'O');

        assertThat(aBoard.getCell(1), is('X'));
        assertThat(aBoard.getCell(9), is('O'));

        Board.Pair[] retrieved = aBoard.get3Cells(1, 5, 9);
        Board.Pair[] expected = {
            new Board.Pair(0, 'X'),
            new Board.Pair(4, '5'),
            new Board.Pair(8, 'O')
        };

        assertThat(retrieved[0], equalTo(expected[0]));
        assertThat(retrieved[1], equalTo(expected[1]));
        assertThat(retrieved[2], equalTo(expected[2]));

        assertThat(aBoard.toString(), not(equalTo(expectedEmptyStr)));
        assertThat(aBoard.toString(), equalTo("X|2|3\n4|5|6\n7|8|O\n"));

        assertFalse(aBoard.isFull());
    }

}
